automates/program_analysis/CAST2GrFN/visitors/cast_to_agraph_visitor.py

```diff
@@ -0,0 +1,449 @@
+    BinaryOperator,
+    Call,
+    ClassDef,
+    Dict,
+    Expr,
+    FunctionDef,
+    List,
+    Loop,
+    ModelBreak,
+    ModelContinue,
+    ModelIf,
+    ModelReturn,
+    Module,
+    Name,
+    Number,
+    Set,
+    String,
+    Subscript,
+    Tuple,
+    UnaryOp,
+    UnaryOperator,
+    VarType,
+    Var,
+)
+
+class CASTToAGraphVisitor(CASTVisitor):
+    """class CASTToAGraphVisitor - A visitor that traverses
+    CAST nodes to generate a networkx DiGraph that represents
+    the CAST as a DiGraph. The CAST object itself is a representation
+    of a program.
+    A common theme across most visitors is they generate a UID
+    that is used with networkx as identifiers for the nodes in the digraph, 
+    so we know which nodes to connect to other nodes with edges. They then
+    add themselves to a networkx DiGraph object that is updated across
+    most the visitors by either adding nodes or edges.
+    A lot of the visitors are relatively straightforward and
+    follow the pattern for a particular node
+        - Visit the node's children
+        - Generate a UID for the current node
+        - Add the node to the graph with the UID
+        - Add edges connecting the node to its children
+        - Return the Node's UID, so this can be repeated as necessary
+
+    Inherits from CASTVisitor to use its visit functions.
+
+    Attributes:
+        cast (CAST): The CAST object representation of the program
+                     we're generating a DiGraph for.
+        G (nx.DiGraph): The graph of the program. Gets populated as 
+                     nodes are visited. 
+
+    """
+    cast: CAST
+    G: nx.DiGraph
+
+    def __init__(self, cast: CAST):
+        self.cast = cast
+        self.G = nx.DiGraph()
+
+    def to_agraph(self):
+        """Visits the entire CAST object to populate the graph G
+        and returns an AGraph of the graph G as a result.
+        """
+        self.visit_list(self.cast.nodes)
+        A = nx.nx_agraph.to_agraph(self.G)
+        A.graph_attr.update(
+            {
+                "dpi": 227, 
+                "fontsize": 20, 
+                "fontname": "Menlo", 
+                "rankdir": "TB"
+            }
+        )
+        A.node_attr.update(
+            {
+                "shape": "rectangle",
+                "color": "#650021",
+                "style": "rounded",    
+                "fontname": "Menlo"
+            }
+        )
+        return A        
+
+    def to_pdf(self,filename):
+        """Creates a pdf of the generated agraph"""
+        A = self.to_agraph()
+        for node in A.iternodes():
+            node.attr["fontcolor"] = "black"
+            node.attr["style"] = "rounded"
+        A.edge_attr.update({"color": "#650021", "arrowsize": 0.5})
+        A.draw(filename+".pdf",prog="dot")
+
+    @singledispatchmethod
+    def visit(self, node: AstNode):
+        """Generic visitor for unimplemented/incorrect nodes"""
+        print("Not implemented: ",node)
```

Comment:
could you make an exception class and raise it here instead?

Reply:
@titomeister to add some context to @dpdicken's request ... it would be useful to have an exception here that stops processing so that we know and can handle cases of unrecognized CAST nodes. Something like the following would be the Pythonic implementation:
```python
def visit(self, node: AstNode):
    """Generic visitor for unimplemented/incorrect nodes"""
    raise TypeError(f"Unrecognized node type: {type(node)}")  
```

We want to show the user the `type` of `node` here because that is the piece of information used by `singledispatch()` to correctly route the call to `visit()`. Also, f-string (`f""`) syntax is very useful for creating strings composed with variables!

Additionally, Daniel's suggestion would help make this even more clear by defining a special error class that would describe the kind of error we would expect here (something like `CASTTypeError`) which could be raised in the place of the `TypeError` in my example.